From b66009fe53e99019170eec23deffdf8225bab1ff Mon Sep 17 00:00:00 2001
From: Gurpreet Ghai <gghai@codeaurora.org>
Date: Fri, 20 Jul 2018 17:25:18 +0530
Subject: [PATCH] BT: Handling aptX Adaptive Mode Change

These changes are for handling aptX Adaptive
mode change between LL and HQ. Also, based on
feedback from app, LE scanning APIs are called
to control the background LE scaning

CRs-Fixed: 2280725

Change-Id: I29e86d1699bc0cb7b875489d56aab306eab9cfee
---
 bta/av/bta_av_api.cc                               |  1 -
 btif/co/bta_av_co.cc                               | 38 ++++++++++++++++++----
 stack/a2dp/a2dp_vendor_aptx_adaptive.cc            |  4 +--
 .../include/a2dp_vendor_aptx_adaptive_constants.h  |  2 +-
 4 files changed, 35 insertions(+), 10 deletions(-)

diff --git a/bta/av/bta_av_api.cc b/bta/av/bta_av_api.cc
index 4bc896a..df8b30b 100644
--- a/bta/av/bta_av_api.cc
+++ b/bta/av/bta_av_api.cc
@@ -384,7 +384,6 @@ void BTA_AvUpdateMaxAVClient(uint8_t max_clients)
   tBTA_AV_MAX_CLIENT *p_buf;
 
   if ((p_buf = (tBTA_AV_MAX_CLIENT *) osi_malloc(sizeof(tBTA_AV_MAX_CLIENT))) != NULL) {
-    APPL_TRACE_DEBUG("%s: gghai: event: %d", __func__, BTA_AV_UPDATE_MAX_AV_CLIENTS_EVT);
     p_buf->hdr.event = BTA_AV_UPDATE_MAX_AV_CLIENTS_EVT;
     p_buf->max_clients = max_clients;
     bta_sys_sendmsg(p_buf);
diff --git a/btif/co/bta_av_co.cc b/btif/co/bta_av_co.cc
index 57bc994..0169925 100644
--- a/btif/co/bta_av_co.cc
+++ b/btif/co/bta_av_co.cc
@@ -76,6 +76,11 @@
 #include "device/include/controller.h"
 #include "bt_vendor_av.h"
 #include "btif/include/btif_storage.h"
+#include <hardware/bt_gatt.h>
+
+#define MAX_2MBPS_AVDTP_MTU 663
+extern const btgatt_interface_t* btif_gatt_get_interface();
+
 bool isDevUiReq = false;
 
 /*****************************************************************************
@@ -1241,7 +1246,11 @@ void bta_av_co_get_peer_params(tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params) {
     min_mtu = p_peer->mtu;
     if (min_mtu > BTA_AV_MAX_A2DP_MTU)
         min_mtu = BTA_AV_MAX_A2DP_MTU;
-    if(min_mtu == 0) min_mtu = 672;
+    if(min_mtu == 0) {
+      APPL_TRACE_WARNING("%s min_mtu received as 0, updating to: %d",
+                               __func__, MAX_2MBPS_AVDTP_MTU);
+      min_mtu = MAX_2MBPS_AVDTP_MTU;
+    }
     APPL_TRACE_DEBUG("%s updating peer MTU to %d for index %d",
                                     __func__, min_mtu, index);
   }
@@ -1297,11 +1306,28 @@ bool bta_av_co_set_codec_user_config(
     goto done;
   }
 
-  if(codec_user_config.codec_type == BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE
-            && codec_user_config.codec_specific_4 > 0) {
-    APPL_TRACE_DEBUG("%s: Updating Encoder Mode to: %x", __func__, codec_user_config.codec_specific_4);
-    BTA_AvUpdateEncoderMode(codec_user_config.codec_specific_4);
-    return success;
+  if(codec_user_config.codec_type == BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE) {
+    if (codec_user_config.codec_specific_4 > 0 ) {
+      const uint16_t ENCODER_MODE_MASK = 0x3000;
+      uint16_t encoder_mode = codec_user_config.codec_specific_4 & ENCODER_MODE_MASK;
+      if (encoder_mode > 0) {
+        APPL_TRACE_DEBUG("%s: Updating Encoder Mode to: %x", __func__, encoder_mode);
+        BTA_AvUpdateEncoderMode(encoder_mode);
+      }
+
+      const uint32_t BLESCAN_MASK = 0x060000;
+      const uint32_t BLESCAN_OFF = 0x020000;
+      const uint32_t BLESCAN_ON = 0x040000;
+      uint32_t blescan_mode = codec_user_config.codec_specific_4 & BLESCAN_MASK;
+      if (blescan_mode & BLESCAN_OFF) {
+        APPL_TRACE_DEBUG("%s: Disabling BLE Scanning", __func__);
+        btif_gatt_get_interface()->scanner->Scan(false);
+      } else if (blescan_mode & BLESCAN_ON) {
+        APPL_TRACE_DEBUG("%s: Enabling BLE Scanning", __func__);
+        //btif_gatt_get_interface()->scanner->Scan(true);
+      }
+      return success;
+    }
   }
 
   // Find the peer SEP codec to use
diff --git a/stack/a2dp/a2dp_vendor_aptx_adaptive.cc b/stack/a2dp/a2dp_vendor_aptx_adaptive.cc
index a469904..82241ac 100644
--- a/stack/a2dp/a2dp_vendor_aptx_adaptive.cc
+++ b/stack/a2dp/a2dp_vendor_aptx_adaptive.cc
@@ -272,7 +272,7 @@ static tA2DP_STATUS A2DP_ParseInfoAptxAdaptive(tA2DP_APTX_ADAPTIVE_CIE* p_ie,
   p_ie->sampleRate = *p_codec_info & 0xF8;
   p_codec_info++;
 
-  p_ie->channelMode = *p_codec_info & 0x0F;
+  p_ie->channelMode = *p_codec_info & 0x1F;
   p_codec_info++;
   //p_ie->channelMode = *(p_codec_info ++);
   p_ie->ttp_ll_0 = *(p_codec_info ++);
@@ -1124,7 +1124,7 @@ bool A2dpCodecConfigAptxAdaptive::setCodecConfig(const uint8_t* p_peer_codec_inf
   LOG_ERROR(LOG_TAG, "setting codec_config_.codec_specific_4 to channelMode: %d", channelMode);
   // Store the channel mode in spare field
   codec_config_.codec_specific_4 &= (int64_t) CHANNEL_MODE_BACK_CHANNEL_MASK;
-  codec_config_.codec_specific_4 |= channelMode;
+  codec_config_.codec_specific_4 |= (int64_t) channelMode << 24;
 
   // Create a local copy of the peer codec capability/config, and the
   // result codec config.
diff --git a/stack/include/a2dp_vendor_aptx_adaptive_constants.h b/stack/include/a2dp_vendor_aptx_adaptive_constants.h
index 8933cce..f75ec04 100644
--- a/stack/include/a2dp_vendor_aptx_adaptive_constants.h
+++ b/stack/include/a2dp_vendor_aptx_adaptive_constants.h
@@ -84,6 +84,6 @@
 #define QCOM_CODEC_APTX_ADAPTIVE_ID                (0xff)
 #define A2DP_APTX_ADAPTIVE_CHANNEL                 (0x0001)
 #define A2DP_APTX_ADAPTIVE_SAMPLERATE              (0x22)
-#define CHANNEL_MODE_BACK_CHANNEL_MASK             (0xE0)
+#define CHANNEL_MODE_BACK_CHANNEL_MASK             (0x1F000000)
 
 #endif  // A2DP_VENDOR_APTX_ADAPTIVE_CONSTANTS_H
-- 
2.7.4

